gnovm/pkg/gnolang/values.go

```diff
@@ -816,8 +816,15 @@ type PackageValue struct {
 	fBlocksMap map[Name]*Block
 }
 
+// IsRealm returns true if pv represents a realm.
+// A user realm, forged by the MsgRun transaction, isn't considered as a realm.
+// XXX The user realm format is:
+// PkgPath="gno.land/r/user_address" PkgName="main"
+// A less confusing format could emerge in the future, with the form of
+// PkgPath="gno.land/u/user_address"
 func (pv *PackageValue) IsRealm() bool {
-	return IsRealmPath(pv.PkgPath)
+	return IsRealmPath(pv.PkgPath) &&
+		pv.PkgName != "main" // Discard user realms forged by MsgRun
```

Comment:
packages can actually be called `main` (there is no restriction on upload) so this creates an attack vector (a realm could call another, while having IsRealm() == false). Can we check the PkgPath to match the format in vmkeeper (which is reserved) instead?

Reply:
Mhh, we could add `/u` straight away, but it's not necessary.

`sdk/vm/keeper.go`:

```
var reRunPath = regexp.MustCompile(`gno\.land/r/g[a-z0-9]+/run`)

// AddPackage adds a package with given fileset.
func (vm *VMKeeper) AddPackage(ctx sdk.Context, msg MsgAddPackage) error {
...
	if reRunPath.MatchString(pkgPath) {
		return ErrInvalidPkgPath("reserved package name: " + pkgPath)
	}
```

so we can use reRunPath in the opposite way (= if it matches, it's not a realm).

Though `/u/` sounds good to, you can re-purpose the PR to do that.

Reply:
Thank you for the details, indeed I can just this regexp, this is the way to go I think. The `/u/` change needs maybe more discussions, I'll leave it for later.
See f103d97